### PR TITLE
[JENKINS-63350] update pac4j to 3.9.0

### PR DESCRIPTION
**Pull request by kuisathaverat**
See [JENKINS-63350](https://issues.jenkins-ci.org/browse/JENKINS-63350).

Update pac4j to 3.9.0 depends on https://github.com/jenkinsci/jenkins/pull/4848

### Submitter checklist

- [X] JIRA issue is well described
- [X] Appropriate autotests or explanation to why this change has no tests

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -78,6 +76,16 @@ under the License.
...
+    <repository>
+      <id>shib-release</id>
+      <url>https://build.shibboleth.net/nexus/content/groups/public</url>
+      <snapshots>
+        <enabled>false</enabled>
+      </snapshots>
+      <releases>
+        <enabled>true</enabled>
+      </releases>
+    </repository>
```
[the remark this answers is not in the document]
Comment (kuisathaverat) [reply pom.xml:88]: I request to mirror the shibboleth repo https://issues.jenkins.io/browse/INFRA-2809

File: pom.xml [hunk 2]
```diff
@@ -7,9 +7,7 @@ Jenkins CI licenses this file to you under the Apache License,
 Version 2.0 (the "License"); you may not use this file except
 in compliance with the License.  You may obtain a copy of the
 License at
-
   http://www.apache.org/licenses/LICENSE-2.0
-
```
[the remark this answers is not in the document]
Comment (kuisathaverat) [reply pom.xml:12]: code formated evil

File: pom.xml [hunk 3]
```diff
@@ -44,7 +42,7 @@ under the License.
   <properties>
     <revision>1.1.7</revision>
     <changelist>-SNAPSHOT</changelist>
-    <jenkins.version>2.176.1</jenkins.version>
+    <jenkins.version>2.251</jenkins.version>
```
[the remark this answers is not in the document]
Comment (kuisathaverat) [reply pom.xml:45]: I am gonna use 2.266

File: pom.xml [hunk 4]
```diff
@@ -98,8 +101,13 @@ under the License.
     <dependency>
       <groupId>org.pac4j</groupId>
       <artifactId>pac4j-saml</artifactId>
-      <version>3.8.3</version>
+      <!-- versions 4.x.x require JDK 11 -->
```
Comment (jglick) [inline pom.xml:104]: Then set `java.level` to `11` I guess? Not sure of status of https://github.com/jenkinsci/plugin-pom/pull/133.

File: Jenkinsfile [hunk 5]
```diff
@@ -1,15 +1,17 @@
...
   [ platform: "linux",   jdk: "8", jenkins: lts, javaLevel: "8" ],
   [ platform: "windows", jdk: "8", jenkins: lts, javaLevel: "8" ],
   [ platform: "linux",   jdk: "11", jenkins: lts, javaLevel: "8" ],
   [ platform: "windows", jdk: "11", jenkins: lts, javaLevel: "8" ],
+*/
   // Also build on recent weekly
-//  [ platform: "linux",   jdk: "11", jenkins: weekly, javaLevel: "8" ],
-//  [ platform: "windows", jdk: "11", jenkins: weekly, javaLevel: "8" ]
+  [ platform: "linux",   jdk: "11", jenkins: weekly, javaLevel: "8" ],
+  [ platform: "windows", jdk: "11", jenkins: weekly, javaLevel: "8" ]
```
Comment (jglick) [inline Jenkinsfile:14]: `javaLevel` should be considered deprecated, and you should not specify `jenkins` unless you _overriding_ the POM version with something newer. Just delete the variables at the top.

```suggestion
  [ platform: "linux",   jdk: "11" ],
  [ platform: "windows", jdk: "11" ]
```

File: pom.xml [hunk 6]
```diff
@@ -98,8 +101,13 @@ under the License.
...
+      <version>3.9.0</version>
       <exclusions>
+        <exclusion>
+          <groupId>org.springframework</groupId>
+          <artifactId>spring-beans</artifactId>
```
Comment (jglick) [inline pom.xml:109]: You may need to exclude more. Pay attention to what gets bundled in the `*.hpi` (also printed to build log).

File: pom.xml (hunk 4 above)
Comment (kuisathaverat) [reply pom.xml:104]: I think is fine, 3.9.0 was released a few days ago

File: pom.xml (hunk 6 above)
Comment (kuisathaverat) [reply pom.xml:110]: yep, there is a bunch of transitive dependencies that can be removed

File: Jenkinsfile [hunk 7]
```diff
@@ -4,14 +4,14 @@ def lts = "2.249.3"
 def weekly = "2.266"
```
Comment (jglick) [inline Jenkinsfile:4]:
```suggestion
```

File: Jenkinsfile [hunk 8]
```diff
@@ -4,14 +4,14 @@ def lts = "2.249.3"
...
+  [ platform: "linux",   jdk: "8", jenkins: lts ],
+  [ platform: "windows", jdk: "8", jenkins: lts ],
+  [ platform: "linux",   jdk: "11", jenkins: lts ],
+  [ platform: "windows", jdk: "11", jenkins: lts ],
 */
   // Also build on recent weekly
-  [ platform: "linux",   jdk: "11", jenkins: weekly, javaLevel: "8" ],
-  [ platform: "windows", jdk: "11", jenkins: weekly, javaLevel: "8" ]
+  [ platform: "linux",   jdk: "11", jenkins: weekly ],
+  [ platform: "windows", jdk: "11", jenkins: weekly ]
```
Comment (jglick) [inline Jenkinsfile:14]: Again, you should let the POM define the baseline unless you are specifically overriding it.

```suggestion
  [ platform: "linux",   jdk: "11" ],
  [ platform: "windows", jdk: "11" ]
```

File: pom.xml (hunk 6 above)
Comment (kuisathaverat) [reply pom.xml:110]: is there any way to acknowledge(remove the warnings) the dependencies? there are some warnings that come from the hpi plugin, but those dependencies are needed and I do not want to add all those dependencies manually to the plugin pom.

Comment (jglick) [reply pom.xml:110]: > is there any way to acknowledge(remove the warnings)

Yes but do not do that.

> there are some warnings that come from the hpi plugin, but those dependencies are needed

You can `exclude` dependencies you know you will not use, or you know you can use in the version supplied by core. Or you can switch to `pluginFirstClassLoader` and forget about accurate testing with `JenkinsRule` (must do it manually). Think twice, actually several times, before doing anything out of the ordinary. https://www.jenkins.io/doc/developer/plugin-development/dependencies-and-class-loading/

Comment (kuisathaverat) [reply pom.xml:110]: I have excluded all that are not needed or are already in the Jenkins Core. so I guess that either I add these libraries to the pom or I live with these warnings

```

[WARNING] Bundling transitive dependency jcommander-1.48.jar (via pac4j-saml)
--
[WARNING] Bundling transitive dependency woodstox-core-5.0.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency istack-commons-runtime-3.0.8.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency FastInfoset-1.2.16.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency metrics-core-3.1.2.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency jakarta.xml.bind-api-2.3.2.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency javax.json-api-1.0.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency joda-time-2.9.9.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency xmlsectool-2.0.0.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency java-support-7.5.0.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency spymemcached-2.12.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency httpclient-4.5.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency httpcore-4.4.8.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency xmlsec-2.1.4.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency velocity-1.7.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency stax2-api-3.1.4.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency cryptacular-1.2.4.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency jaxb-runtime-2.3.2.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency txw2-2.3.2.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency javax.json-1.0.4.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency stax-ex-1.8.1.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency ldaptive-1.0.13.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-core-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-messaging-api-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-messaging-impl-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-profile-api-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-profile-impl-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-saml-api-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-saml-impl-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-security-api-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-security-impl-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-soap-api-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-soap-impl-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-storage-api-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-storage-impl-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-xmlsec-api-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency opensaml-xmlsec-impl-3.4.3.jar (via pac4j-saml)
[WARNING] Bundling transitive dependency pac4j-core-3.9.0.jar (via pac4j-saml)
```

Comment (kuisathaverat) [reply pom.xml:110]: I will live with the warnings to remove them I have to add more than 300 no-sense lines to the pom to manage these dependencies manually.

**Comment by kuisathaverat** [comment]
I have manually tested it, it works. There is a small detail on the URL for the SAML response I have to convert into optional in some way but it works. This is my test environment https://github.com/kuisathaverat/jenkins-issues/tree/master/JENKINS-63350

**Comment by kuisathaverat** [comment]
resolved, the plugin works as works the previous version now.

File: pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (jglick) [reply pom.xml:91]: https://wiki.shibboleth.net/confluence/display/DEV/Use+of+Maven+Central

**Review by amuniz — APPROVED** [review]
[INFRA-2809](https://issues.jenkins.io/browse/INFRA-2809) is resolved and the plugin builds fine locally.

I think this can be merged now @kuisathaverat - thanks for the work here!

**Comment by amuniz** [comment]
Actually, it seems to be downloading the dependency from the shibboleth repo. It's not listed in the [Jenkins one](https://repo.jenkins-ci.org/list/shibboleth/).

**Comment by amuniz** [comment]
OK, all artifacts are now available in the Jenkins repo, so all good.

File: pom.xml [hunk 9]
```diff
@@ -78,6 +78,17 @@ under the License.
...
+    <repository>
+      <id>shib-release</id>
+      <url>https://build.shibboleth.net/nexus/content/groups/public</url>
+      <snapshots>
+        <enabled>false</enabled>
+      </snapshots>
+      <releases>
+        <enabled>true</enabled>
+      </releases>
+    </repository>
```
Comment (kuisathaverat) [inline pom.xml:91]:
```suggestion
```

**Comment by kuisathaverat** [comment]
I will wait for the CI to end to merge the PR

**Comment by timja** [comment]
finished ^^

**Comment by kuisathaverat** [comment]
I will release on the weekend

File: Jenkinsfile [hunk 10]
```diff
@@ -1,15 +1,16 @@
 //def buildConfiguration = buildPlugin.recommendedConfigurations()
 
-def lts = "2.176.1"
-def weekly = "2.199"
+def lts = "2.249.3"
```
Comment (jglick) [inline Jenkinsfile:3]: ??

File: src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java [hunk 11]
```diff
@@ -39,8 +39,8 @@
 import org.jenkinsci.plugins.saml.user.SamlCustomProperty;
 import org.kohsuke.stapler.*;
 import org.kohsuke.stapler.interceptor.RequirePOST;
-import org.pac4j.core.client.RedirectAction;
-import org.pac4j.core.client.RedirectAction.RedirectType;
+import org.pac4j.core.redirect.RedirectAction;
+import org.pac4j.core.redirect.RedirectAction.RedirectType;
 import org.springframework.dao.DataAccessException;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:44]: Do you not plan to switch to Spring Security versions of Jenkins interfaces as well, now that you are on a baseline with JEP-227?

Comment (kuisathaverat) [reply src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:44]: sure, what is the replacement for DataAccessException?

Comment (jglick) [reply src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java:44]: There is none.

Best to look at other PRs associated with JEP-227 for examples.

**Pull request merged**